### ISSUE TITLE
feat(gpu): HIP backend cannot run qwen3_5_moe (35B-A3B) — fused QKV + GatedDeltaNet + MoE not supported

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#1831]
## Summary

The HIP 1BP GPU backend (`src/backend_hip_1bp.cpp`) rejects all `qwen3_5_moe` / `qwen35moe` models at init with `unsupported architecture: missing required tensors (attn_v/ffn_gate/ffn_up/ffn_down). Hybrid SSM/MoE models are not supported by the GPU backend`. This blocks the 35B-A3B (and any GatedDeltaNet-MoE model) from the GPU path — the only GPU-capable dense-GQA path is `hip_1bp`, which is hard-coded for dense transformers.

## What the 35B needs that the GPU backend lacks

From the official `Qwen3.6-35B-A3B-NPU2/config.json` (model_type `qwen3_5_moe`) and the GGUF tensor list:

1. **Fused QKV**: tensors are `blk.N.attn_qkv.weight` (fused), not separate `attn_q/attn_k/attn_v` — the backend's `gt()` looks up separate Q/K/V names.
2. **GatedDeltaNet linear attention** (40 layers, 30 linear-attn + 10 full-attn per config `full_attention_interval: 4`): tensors `ssm_a`, `ssm_alpha`, `ssm_beta`, `ssm_conv1d`, `ssm_dt.bias`, `ssm_norm`, `ssm_out`, `attn_gate`.
3. **MoE experts**: `ffn_gate_exps/ffn_up_exps/ffn_down_exps.weight` (256 experts, top-8) + `ffn_gate_inp` (router) + `ffn_gate_shexp/up_shexp/down_shexp` (shared experts) — the backend looks for dense `ffn_gate/up/down` only.
4. **attn_output_gate: true** in config (fused output gating).

## Impact

- `1bit zaya --model model.q4nx` on a 35B falls back to CPU (or fails) because HIP 1BP rejects the arch, and no other GPU backend handles `qwen3_5_moe`.
- The community benchmarks this exact model at 100+ tok/s on GPU (ROCmFP4 fork + MTP); the engine's GPU path cannot run it at all.
- The engine DOES run this arch on NPU (`npu_engine_universal` — see #1830 for its separate decode bug) and CPU (`src/qwen3next_engine.cpp`, a validated reference: GatedDeltaNet + GQA + gated MoE, corr 0.9997 vs numpy).

## Suggestion

`src/qwen3next_engine.cpp` is a validated CPU reference for the exact arch. Porting its layer dispatch (linear_attention | full_attention, MoE softmax router + norm_topk + gated experts) into the HIP packed-kernel framework would unlock the 35B on GPU. This is the architecture behind the community's headline Strix Halo numbers.

## Hardware

- AMD Ryzen AI MAX+ 395 (Strix Halo), 128 GB unified, ROCm TheRock 7.16, gfx1151
- Reproduced with both the community `Qwen3.6-35B-A3B-Q4_0_ROCMFP4_STRIX_LEAN.gguf` (19 GB) and the official NPU `model.q4nx`

**Comment by bong-water-water-bong** [comment]
**Owner status (2026-09-02): scoped — the GDN math has an NPU reference to mirror; the HIP lane is a defined feature.**

The arch is registry-known (RCPP_ARCH_QWEN35 = Qwen3.5 Gate-Delta Net: fused QKV + SSM path + GDN attention; loader hint already routes it to NPU/HIP). Two reference implementations exist to port the math from, so this is an implementation task, not reverse-engineering:

1. **engine/npu universal path** (`npu_engine_universal.cpp` + the npu-infer 35B decode work on feat/hrx) — Qwen3.6-35B-A3B fused-QKV + GatedDeltaNet tensors (`attn_qkv`, `ssm_a`, `ssm_b`, `ssm_dt`, gate/up/down MoE) load + run on the NPU side.
2. The 35B weight-BO/desc layout work (npu-infer docs/35b-*, dump_moe_layout) decoded the tensor family the GPU path would consume.

What backend_hip_1bp needs (unchanged from the issue): fused-QKV name handling in gt() + a GatedDeltaNet linear-attention kernel path (the full_attention_interval=4 hybrid schedule). No HIP GDN kernels exist yet in src/ (only dense-GQA + bonsai gemv families). Open; next owner step = port the GDN attention from the NPU-universal reference into backend_hip_1bp behind the qwen35 arch gate.

**Comment by bong-water-water-bong** [comment]
**M1 landed — port map (2026-09-05).** PR #2121 adds `docs/research/hip-qwen35-gdn-port.md`: arch-gate ground truth (`RCPP_ARCH_QWEN35=21`, GGUF strings interns2preview/kimik2/quark/qwen3_5dllm), the validated CPU math reference (qwen3next_engine.cpp), the NPU schema family (npu_engine_universal `model.layer.N.linear_attn.ssm_*`/`mlp.*_exps_proj`), and M1–M5 with the strixhalo corr acceptance gate.

**Why M1 is plan-first:** the community GGUF's exact kv-keys/tensor naming (fused attn_qkv row layout, linear-head dims, layer_types encoding) is confirmed in NO checked-out source — the producing converter is in the ROCmFP4/llama.cpp fork. A loader written against guessed keys would silently find nothing (the #1624/#1627 failure mode). 

**Blockers to lift for M2 (code):** (1) schema capture of the target 35B file — H1BP_SCHEMA_DUMP debug env (kv_keys + tensor table + fused-QKV slice geometry), run against a real 35B GGUF/q4nx on strixhalo; (2) router HIP-candidate wiring for qwen35moe; (3) a 35B model staged at ~/models on strixhalo (current set: GLM-4.7/MiniCPM/Qwen3-0.6B/zaya + qwen3next-80b q4nx). Next owner: GPU/HIP lane — needs one of the two target files staged to proceed.

**Comment by bong-water-water-bong** [comment]
**M2 validated on real hardware (2026-09-05).** The qwen35moe structural loader now runs on strixhalo (gfx1151, ROCm TheRock) against the real 36.9GB `Qwen3.6-35B-A3B-Q8_0.gguf` (downloaded + staged at ~/models):

```
[hip1bp] qwen35moe structure verified (H=2048 L=40, 10 shared + kind-specific per layer;
10 full-attn MHA layers 3,7,…,39 with split attn_q/k/v + q/k-norm, rest GDN fused
attn_qkv 8192 rows + ssm; MoE 256x8 + shared): GDN/MoE decode kernels not implemented
yet (#1831 M3/M4)
```

First-run caught a real schema bug (assumed uniform 19 tensors/layer; full-attn layers 3,7,…,39 are 16-tensor split MHA) — fixed, and the hardware-confirmed two-kind schema is recorded in the port map (`docs/research/hip-qwen35-gdn-port.md` §captured-schema). The loader was also proven to catch cfg/discovery mismatches (bench-default dims flagged, schema still verified on shapes). PR #2121 carries M1+M2 (docs + loader + bench GGUF-tolerance driver). Next: M3 — GDN kernel math (attn_qkv slice semantics etc., open questions in the port map), validated against the same staged file.

**Comment by bong-water-water-bong** [comment]
**M2 merged — PR #2121 on main (932ed5ba, 2026-09-05 17:48Z).** Branch cleanup done (remote branch deleted; strixhalo validation worktree retained for M3). Next milestone M3: GDN kernel math — attn_qkv 8192-row slice semantics, attn_gate 4096-row role, ssm group/head mapping, fused-expert indexing (open questions in docs/research/hip-qwen35-gdn-port.md §captured-schema).

**Comment by bong-water-water-bong** [comment]
**M3 milestone (2026-09-05) — decode validated vs in-repo golden.** Summary for #1831:

**What landed on feat/hip-qwen35-m3-decode (PR #2127):** eager qwen35moe GPU decode — embed → 40 layers (10 full-attn MHA w/ IMRoPE-as-nrot64 + q/gate split; 30 GDN fused-qkv + conv + delta rule; MoE top-8 + shared expert every layer) → output_norm → lm_head (Q8_0 raw gemvs throughout). Env-gated (H1BP_Q35_LOAD + H1BP_Q35_TRY), so no default-path change.

**Validation:** engine hidden states == host formula **corr 1.00000 over all 40 layers** (single token) and **corr 0.99884 with argmax-equality** on a stateful 2-token decode (cross-token conv/state/KV handling covered). Bugs found+fixed en route: stream-race D2D copies, delta-kernel state-stride overlap, full-attn q/gate de-interleave.

**Residual risks (honest):** (1) formula-vs-llama-HIP final-logits content corr ~0.1 at pos0 despite identical magnitudes + byte-identical weights — a small semantic divergence from llama's decode remains unlocalized (llama CPU qwen35moe is broken upstream: fused-path heap corruption + degenerate NOFUSE output, so no clean per-layer oracle); (2) validation used 2-token golden (full 100-token logits-corr not runnable — golden cost); text-level greedy vs llama not yet compared; (3) 1BP q4nx + packed-quant fast paths + perf remain out of scope (as planned). Recommended follow-ups: file the llama.cpp qwen35moe CPU defects upstream; 100+-token engine-vs-golden run; then remove the env gates once corr/consistency targets hold.

**Comment by bong-water-water-bong** [comment]
**Formula divergence root-caused and validated (2026-09-05 late).** Two bugs separated the HIP engine/formula from llama:
1. **GDN q/k L2-norm form**: llama ggml_l2_norm is SUM-form (unit rows); the engine used RMS/mean form leaving ||k||=√128 → delta-rule instability (state inf by token ~25) + 11.3x-scaled q/k. Fix = sum-form (PR #2134).
2. **MoE norm_topk renorm** (llama sum+clamp of selected probs) was off by default. On = matches llama.

Result: in-repo CPU reference (PR #2131) vs llama-HIP healthy decode over a 99-token sequence — per-position logits corr **min 0.9937 / mean 0.9984, all ≥ 0.99** (residual = llama BF16 noise). The engine implements this formula (engine==formula verified 40/40 layers earlier); the two one-line fixes are in PR #2134, pending CI + strixhalo hardware re-verify (strixhalo HIP toolchain regressed tonight — clang 23 cuda_wrappers/new — blocking local rebuilds). llama-CPU qwen35moe defects (heap corruption + degenerate loops) remain upstream-blocking for the CPU oracle.

**Comment by bong-water-water-bong** [comment]
**M3 hardware validation COMPLETE (2026-09-06).** The two formula fixes (GDN q/k L2-norm sum-form; MoE norm_topk renorm default ON) are in PR #2134; engine rebuilt locally at -std=c++23 (workaround for the clang-23/libstdc++ c++26 HIP toolchain regression — repo-wide, pending a build-config decision) and validated over a 99-token sequence vs the in-repo CPU reference and llama-HIP:

- ENGINE vs CPU-ref: corr min 0.99555 / mean 0.99869 — **all positions ≥ 0.99**
- CPU-ref vs llama-HIP (healthy decode): min 0.99368 / mean 0.99842 — all ≥ 0.99
- ENGINE vs llama-HIP: min 0.98982 / mean 0.99793 (98/99 ≥ 0.99; residual = llama BF16 compute noise)

**Result:** qwen35moe HIP decode matches the reference at ≥ 0.99 over ~100 tokens — the objective's gate, with the in-repo CPU reference (PR #2131) standing in for llama CPU (broken upstream for qwen35moe: heap corruption + degenerate output — documented). Remaining: merge PRs #2131/#2134 once the repo-wide HIP toolchain regression (clang-23 cuda_wrappers/new under C++26; all HIP CI fails) is resolved via a build-config decision; then the merged engine re-verifies on the normal toolchain.

**Comment by bong-water-water-bong** [comment]
**All M3 PRs merged to main** (#2135 infra C++23 toolchain fix, #2131 CPU reference, #2134 engine L2-norm + norm_topk fixes). Final state: qwen35moe HIP decode validated ≥ 0.99 corr over ~100 tokens vs the in-repo CPU reference (which itself matches llama-HIP at ≥ 0.99), after root-causing the GDN L2-norm form and MoE renorm bugs. Remaining follow-ups (out of scope, tracked): 1BP q4nx + packed-quant fast paths, perf, hipGraph capture, and upstream llama.cpp qwen35moe CPU defects. #1831 stays open for those.

**Comment by bong-water-water-bong** [comment]
## 1BP q4nx lane — milestone 1: conversion works + verified (residual 1 progress)

The qwen35moe GGUF → 1BP q4nx path is now proven end-to-end on strixhalo:

- **Converted**: `Qwen3.6-35B-A3B-Q8_0.gguf` (36.9 GB) → `/var/tmp/cas/qwen35-1bp.1bp` (20.7 GB q4nx) via `gguf_to_onebp --q4nx` (src/gguf_to_onebp.cpp's qwen35moe branch → ONEBP_MOE). 156 s.
- **Full 733-tensor routing**: incl. the GDN family (fused `attn_qkv` 8192×2048, `attn_gate`, `ssm_conv1d`, ssm_* per layer), the split-MHA layers (3,7,…,39: attn_q/k/v + norms), MoE 256-exps + shared experts, router, norms — the captured schema is respected.
- **Verified PASS**: `verify_onebp` over all 733 tensors — Q4NX maxabs within quant bounds (global 6.7e-2, per-tensor ≤ bound), 0 aliased.

**Next (milestone 2, the actual lane deliverable)**: wire `backend_hip_1bp.cpp`'s qwen35 block to consume the 1BP q4nx model — populate the Q35L per-layer structs from `NpuOnebpModel::get_tile_ptr` (the dense path's conventions, launch_q4nx-style tile dequant) instead of the GGUF-direct Q8_0 raw loads, behind the H1BP_Q35_* gates; gate = the qwen35moe_cpu_ref corr methodology. The converted file is staged on strixhalo (`/var/tmp/cas/qwen35-1bp.1bp`) — it will be re-staged to ~/models when milestone 2 starts.

**Comment by bong-water-water-bong** [comment]
## 1BP q4nx lane — milestone 2 COMPLETE (PR #2138)

`backend_hip_1bp.cpp`'s qwen35 block now consumes the 1BP **Q4NX** model: Q35L per-layer structs populated from `NpuOnebpModel` tiles with `launch_q4nx`-style gemvs (incl. per-expert byte-offset slices of the ndim==3 stacks), behind the existing `H1BP_Q35_LOAD`/`H1BP_Q35_TRY` gates. Zero new kernels; GGUF-direct path untouched. Supporting changes: qwen35 detection probes the 1BP tensor index (no GGUF arch token in 1BP headers); `gguf_to_onebp --q4nx` gained per-tensor F16 routing (token_embd/output, MoE router, ssm conv1d — lossless where coarse codebooks corrupt); new `tools/qwen35moe_cpu_ref_1bp` (1BP fork of the #2131 reference) enables a **same-file** corr gate.

**Validation (strixhalo gfx1151, 100-token sequence, identical inputs):**
- Same-file gate (engine-on-q4nx vs 1BP cpu-ref): corr **mean 0.9983 / min 0.9791 @pos56 / argmax-eq 99/100 / pos0-2 = 1.0000 (bit-identical)** — statistically equal to the M3 GGUF baseline measured on the same sequence (mean 0.9982 / pos0-3 = 1.0000 / same ~0.98 min floor @pos56 = engine-vs-ref accumulation-order residual, lane-independent).
- Structural: 22.7–24.0 GB loaded across 40 layers + globals [1BP Q4NX]; eager decode 28 tok/s.
- Cross-file note: vs the Q8_0 GGUF oracle the all-Q4NX artifact lands at mean 0.9754 — pure quantization delta of per-layer weights (v1 0.9690 → v2 0.9754 after F16 routing); wiring faithfulness proven by the oracle-invariant Q4NX-vs-GGUF-engine curve and the same-file 1.0s.
Evidence: docs/research/hip-qwen35-m2-correlation.txt (PR #2138), docs/research/hip-qwen35-1bp-q4nx-wiring.md.

**Remaining follow-ups (out of scope here, tracked on #1831):** ROCmFP4/F16/TQ2NZ 1BP quants for qwen35, perf (packed fast paths, hipGraph capture for q35), model_discovery v4-header/arch-enum modernization, upstream llama.cpp qwen35moe CPU defects.

**Comment by bong-water-water-bong** [comment]
## ✅ Resolved — qwen35moe runs on the HIP GPU backend (closing)

The original gap ("HIP 1BP backend rejects every qwen3_5_moe model") is fixed and merged on main:

- **M1/M2/M3 — GGUF-direct decode** (PRs #2121, #2127, #2131, #2134, #2135): structural loader for the captured 733-tensor schema + eager GDN/MoE decode from Q8_0 GGUF; formula fixes (GDN q/k L2-norm sum-form, MoE norm_topk renorm); ≥0.99 per-position logits corr vs the in-repo CPU reference.
- **1BP q4nx lane M1/M2** (PR #2138, merged `f7b4df155709`): `backend_hip_1bp.cpp` consumes 1BP Q4NX models through the same Q35L structs + `launch_q4nx` dispatch (per-expert slices), `gguf_to_onebp --q4nx` per-tensor F16 routing (token_embd/output/router/conv1d), and a same-file CPU oracle (`tools/qwen35moe_cpu_ref_1bp`): engine-on-q4nx vs q4nx-cpu-ref over 100 tokens = corr mean 0.9983 / argmax-eq 99/100 / pos0-2 bit-identical — equal to the M3 GGUF baseline on the same sequence.
- Evidence: `docs/research/hip-qwen35-m2-correlation.txt`, `docs/research/hip-qwen35-1bp-q4nx-wiring.md`, `docs/research/hip-qwen35-gdn-port.md`.

**Tracked as future work (separate issues when started — none block GPU use of the 35B):**
- Other 1BP quants for qwen35 (ROCmFP4 / F16 / TQ2NZ files) on the hip_1bp lane
- Perf: packed fast paths per tensor class + hipGraph capture for the q35 decode
- model_discovery v4-header/arch-enum modernization (routing nicety)
- Upstream llama.cpp qwen35moe CPU defects (heap corruption / degenerate output) — external

Hardware validation: strixhalo (gfx1151, TheRock); dense-model behavior unchanged (env-gated lanes).
